documentation.md

```diff
@@ -0,0 +1,97 @@
+case, this means taking advantage of [Django Channels](https://github.com/django/channels).
+
+The Django Streams integration is heavily inspired by the helpers and mixins found in
+[turbo-rails](https://github.com/hotwired/turbo-rails). There are a few steps required to get
+the integration working.
+
+### Initial Setup
+#### Django Channels
+After setting up Channels according to the [documentation](https://channels.readthedocs.io/en/stable/installation.html),
+make sure to modify the top-level `ProtocolTypeRouter` to include the `TurboStreamsConsumer`:
+
+```python
+from turbo.consumers import TurboStreamsConsumer
+application = ProtocolTypeRouter({
+  "http": AsgiHandler(),
+  "websocket": TurboStreamsConsumer.as_asgi()  # Leave off .as_asgi() if using Channels 2.x
+})
+```
+
+#### Including JavaScript
+In our `<head>` for all our templates, we need to `{% include 'turbo/head.html' %}`, which will include the Turbo libraries we need
+in order to use Turbo Drive, Frames and Streams in our app.
+
+### Model Layer
+
+Throughout these docs we'll be using these two basic models for a chat app, similar to the
+[Hotwire demo video](https://www.youtube.com/watch?v=eKY-QES1XQQ):
+
+```python
+# chat/models.py
+
+from django.db import models
+class Room(models.Model):
+    name = models.CharField(max_length=255)
+
+
+class Message(models.Model):
+    room = models.ForeignKey(Room, related_name="messages", on_delete=models.CASCADE)
+    text = models.TextField()
+```
+
+`turbo-django` provides a `BroadcastableMixin` that enables CRUD operations from a given model
+to be broadcast over websockets. To enable messages to be broadcast over websocket to users
+listening in on a particular room, we add the `BroadcastableMixin` and tell `turbo-django`
+that actions on messages should be broadcast to the related `Room` as well:
+
+```python
+class Message(BroadcastableMixin, models.Model):
+    broadcast_to = ["room"]
+    broadcast_self = False
+
+    room = models.ForeignKey(Room, related_name="messages", on_delete=models.CASCADE)
+    text = models.TextField()
+    created_at = models.DateTimeField(auto_now_add=True)
+```
+
+`broadcast_to` can contain any foreign key field on the model. We set `broadcast_self` to `False` since we don't
+anticipate listening in to updates on a single message, only on the associated `Room`. This is all the set-up on the
+model side of things that we'll need!
+
+### Template Layer
+
+#### Subscribing to a stream
+Next, we can declare that a given page should subscribe to broadcasts for a given model instance. Make sure to
+`{% load turbo_streams_helpers %}` to get access to the template tag in your template.
+
+Make sure to pass the model instance (here, it's passed as `room`) into the template context. Somewhere in the `<body>`,
+we need to add the template tag `{% turbo_stream_from room %}`. That's all we need to do to connect this template, when
```

Comment:
OK, I'm back from reading this and the Rails code. I have a few observations from the Rails code that helped me understand this better. 

The first is that the Rails code seems to give more control to the dev around what gets sent to the stream by keeping the stream name pretty separate from the models. The second is letting the dev send updates to the stream themselves. I like the Rails way more since it allows for more flexibility and is a bit more explicit about what's being sent to the front end. 

A rough translation for the template tag of the Rails way might look like this: `{% turbo_stream_from room "messages" %}`, which would roughly output a channel name of `TurboStream::Room:1:messages`, where `1` is the room pk.

Then in the Messages model instead of `broadcast_to` we might have a method that the mixin sets up for the user to override that's called after the `save` has completed successfully. Here the method is called `broadcast`:

```python
class Message(BroadcastableMixin, models.Model):
    room = models.ForeignKey(Room, related_name="messages", on_delete=models.CASCADE)
    text = models.TextField()
    def broadcast(self):
        self.broadcast_append_to(self.room, "messages")
```

The full signature for `broadcast_append_to` might look like `broadcast_append_to(model, target=None, partial=None, context=None)`. Where `partial` would default to `app_name/model_name.html` and `context` would default to `{"object": self}`.

The neat thing about this is the user can now filter or modify what's sent to the front end. Say Message has a `private` field that we don't want to broadcast to some streams:

```python
class Message(BroadcastableMixin, models.Model):
    ...
    def broadcast(self):
        context = {"object": self, "public": True}
        if not self.private:
            self.broadcast_append_to(self.room, "messages_public", context=context)
        self.broadcast_append_to(self.room, "messages")
```

Along with `broadcast_append_to` Rails has `broadcast_prepend_to` and such, see: https://github.com/hotwired/turbo-rails/blob/main/app/models/concerns/turbo/broadcastable.rb

Anyway, don't mean to derail this PR, but just wanted to see what you thought about this. Cheers!

Reply:
Thanks for taking the time to parse through the Ruby code over there! Definitely something I need to work on. I think a lot of this makes sense I like the idea of `broadcast` being a method that can be overridden per-model for customizations. I still feel like there's a place for a default implementation which simply iterates through `broadcasts_to` and calls `broadcast_[append|prepend]_to` with the proper arguments.

I'm wondering how splitting up channels by property should work. It does make sense that if you subscribe to a `Room` where both `Room` and `Message` are subscribable, there's instances where you may only want `messages` or only want the `room` itself. If you wanted to subscribe to two properties, what would the syntax look like? Would you have two `turbo_stream_from` tags? If you want to subscribe to updates on the model itself (not on an associated many-to-one), would you write `{% turbo_stream_from room "self" %}`? Or should `self` always be included?

I'll work on implementing the first part of the feedback with the `broadcast` override, but I'll leave the second part open so we can continue discussion.

Reply:
On some further inspection, I'm not sure how much `broadcast_append_to` and `broadcast_prepend_to` need to be separate functions -- it seems like the only difference is whether the action is `append` or `prepend`, which seems to be much better suited to be an argument to a generic function.

Reply:
I like the discussion and there are many good points brought up by @blopker!
Although i would suggest that we split some of the comments into new issues and develop them iteratively. 

Regarding the room and message discussion i personally see it more about general topics and the model and message as a special case of that. 

So i would keep the topic general and probably have convenience methods to generate a normalized topic from a model. Wdyt?

Reply:
Check out issue #14 for discussion on generalizing streams/subscriptions beyond models :)